[Merged by Bors] - feat(FieldTheory.JacobsonNoether) : add proof of the Jacobson-Noether theorem

faenuccio opened the pull request:
In this PR we provide a proof of the Jacobson-Noether theorem following [this blog](https://ysharifi.wordpress.com/2011/09/30/the-jacobson-noether-theorem/)

Co-authored by: Filippo A. E. Nuccio @faenuccio
Co-authored by: Wanyi He @Blackfeather007
Co-authored by: Huanyu Zheng @Yu-Misaka
Co-authored by: Yi Yuan @yuanyi-350
Co-authored by: Weichen Jiao @AlbertJ-314
Co-authored by: Sihan Wu @Old-Turtledove
Co-authored by: @FR-vdash-bot

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:



If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

acmepjz commented:
There is API for purely inseparable extension (unfortunately, only for commutative rings). Would you consider using it (after extend it to non-commutative case)?

Yu-Misaka commented:
> There is API for purely inseparable extension (unfortunately, only for commutative rings). Would you consider using it (after extend it to non-commutative case)?

I think indeed we've tried using `[isPurelyInseparable_iff_pow_mem](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/PurelyInseparable.html#isPurelyInseparable_iff_pow_mem)`. But like you said, we need to either extend purely inseparable to non-commutative case or figure out how to define `k(\alpha)` in the notation of that blog (the difficulty here is that `IntermediateField` demands a field over `k(\alpha)`, which isn't easy to find). This is surely an interesting idea.

Yu-Misaka reviewed (commented):
Sorry I didn't notice the change of `open Classical` in @Blackfeather007 's commit. It is indeed better to be placed behind.

Whysoserioushah commented:
I think instead of `open Classical`, the right thing to do is use tactic `Classical` whenever you need it

acmepjz commented:
Do you think this is a better statement of the result?
```lean
import Mathlib.RingTheory.Algebraic
import Mathlib.FieldTheory.Separable

example {k D : Type*} [Field k] [DivisionRing D] [Algebra k D]
    [Algebra.IsAlgebraic k D]
    (hneq : (⊥ : Subalgebra k D) ≠ ⊤) :
    ∃ x : D, x ∉ (⊥ : Subalgebra k D) ∧ IsSeparable k x := by
  sorry
```
Note that this is not necessary correct; I haven't assume `k` is the center of `D`.

The advantage of this statement is that now `k` and `D` are both types, while your approach `k` is a subring, then converted to a type.

Whysoserioushah commented:
> Do you think this is a better statement of the result?
> 
> ```lean
> import Mathlib.RingTheory.Algebraic
> import Mathlib.FieldTheory.Separable
> 
> example {k D : Type*} [Field k] [DivisionRing D] [Algebra k D]
>     [Algebra.IsAlgebraic k D]
>     (hcenter : Subalgebra.center k D = ⊥)
>     (hneq : (⊥ : Subalgebra k D) ≠ ⊤) :
>     ∃ x : D, x ∉ (⊥ : Subalgebra k D) ∧ IsSeparable k x := by
>   sorry
> ```
> 
> ~Note that this is not necessary correct; I haven't assume `k` is the center of `D`.~ The center assumption is needed, otherwise a couhterexample is `D` a (commutative) purely inseparable extension of `k`. But what about this: change the assumption to `Subalgebra.center k D ≠ ⊤` and the result to `∃ x : D, x ∉ Subalgebra.center k D ∧ IsSeparable k x`?
> 
> The advantage of this statement is that now `k` and `D` are both types, while your approach `k` is a subring, then converted to a type.

Then probly the assumption `hneq` is redundant since if ⊥ = ⊤ then not in ⊥ is in the empty set

Whysoserioushah commented:
Also it's probly not the best idea to use `rename_i`

faenuccio commented:
> Do you think this is a better statement of the result?
> 
> ```lean
> import Mathlib.RingTheory.Algebraic
> import Mathlib.FieldTheory.Separable
> 
> example {k D : Type*} [Field k] [DivisionRing D] [Algebra k D]
>     [Algebra.IsAlgebraic k D]
>     (hcenter : Subalgebra.center k D = ⊥)
>     (hneq : (⊥ : Subalgebra k D) ≠ ⊤) :
>     ∃ x : D, x ∉ (⊥ : Subalgebra k D) ∧ IsSeparable k x := by
>   sorry
> ```
> 
> ~Note that this is not necessary correct; I haven't assume `k` is the center of `D`.~ The center assumption is needed, otherwise a couhterexample is `D` a (commutative) purely inseparable extension of `k`. But what about this: change the assumption to `Subalgebra.center k D ≠ ⊤` and the result to `∃ x : D, x ∉ Subalgebra.center k D ∧ IsSeparable k x`?
> 
> The advantage of this statement is that now `k` and `D` are both types, while your approach `k` is a subring, then converted to a type.

This is probably a good point; I would suggest that we try to keep *both* versions, provided the proof of the second is very very short using the first one.

faenuccio commented:
@Whysoserioushah I have seen your commit e86ef6420621ebc0b9f836de5958c2ee15e9691e and that you have removed some comment I made. It is normally a good habit to only leave the authors of one branch to commit/push, and to add suggestions that they can point-by-point decide to accept or reject. By the way, do you have a Zulip account, so that we can communicate more directly there (which is certainly more appropriate than here)? Thanks!

acmepjz commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,258 @@
+
+instance : Algebra (Subring.center D) D := Algebra.ofModule smul_mul_assoc mul_smul_comm
+
+private def f (a : D) : D →ₗ[k] D := {
+  toFun := fun x ↦ a * x
+  map_add' := fun x y ↦ LeftDistribClass.left_distrib a x y
+  map_smul' := fun m x ↦ by simp only [mul_smul_comm, RingHom.id_apply]
+}
```
Maybe you can use [LinearMap.mulLeft](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Algebra/Bilinear.html#LinearMap.mulLeft) for `f`, [LinearMap.mulRight](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Algebra/Bilinear.html#LinearMap.mulRight) for `g`, [LinearMap.commute_mulLeft_right](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Algebra/Bilinear.html#LinearMap.commute_mulLeft_right) for they commutes, etc. etc.

Yu-Misaka replied:
good idea, implemented, thanks!

acmepjz commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,258 @@
+@[simp]
+private lemma g_def (a x : D) : g a x = x * a := rfl
+
+private def δ (a : D) : D →ₗ[k] D := {
+  toFun := f a - g a
+  map_add' := fun x y ↦ by simp only [Pi.sub_apply, map_add, map_add, add_sub_add_comm]
+  map_smul' := fun m x ↦ by simp only [Pi.sub_apply, map_smul, RingHom.id_apply, smul_sub]
+}
```
Maybe you can just write `private def δ (a : D) : D →ₗ[k] D := f - g`

Yu-Misaka replied:
implemented

Whysoserioushah commented:
> @Whysoserioushah I have seen your commit [e86ef64](https://github.com/leanprover-community/mathlib4/commit/e86ef6420621ebc0b9f836de5958c2ee15e9691e) and that you have removed some comment I made. It is normally a good habit to only leave the authors of one branch to commit/push, and to add suggestions that they can point-by-point decide to accept or reject. By the way, do you have a Zulip account, so that we can communicate more directly there (which is certainly more appropriate than here)? Thanks!

Yes sorry I thought they consent me to edit on their branch when we were in PKU, my zulipchat account name is Edison Xie, id is 699016 we should probably make a groupchat in case of further possible project for them what do you think? :)

faenuccio commented:
> > @Whysoserioushah I have seen your commit [e86ef64](https://github.com/leanprover-community/mathlib4/commit/e86ef6420621ebc0b9f836de5958c2ee15e9691e) and that you have removed some comment I made. It is normally a good habit to only leave the authors of one branch to commit/push, and to add suggestions that they can point-by-point decide to accept or reject. By the way, do you have a Zulip account, so that we can communicate more directly there (which is certainly more appropriate than here)? Thanks!
> 
> Yes sorry I thought they consent me to edit on their branch when we were in PKU, my zulipchat account name is Edison Xie, id is 699016 we should probably make a groupchat in case of further possible project for them what do you think? :)

OK, no problem: and thanks for the zulip name! Concerning the group chat, our hope is that everybody will be a regular active Zulip user, and private groups go a bit against this philosophy. I normally use DM with groups that I mentor, as in this case, but then the public chat is preferable.

acmepjz commented:
By the way, out of curiosity I'm working on a much shorter proof (also based on the blog link you posted), which is expected to be ~100 lines. I think I should not spoil you for now 😂

Yu-Misaka reviewed (commented):
This might not be a good commit and a recompilation is needed.
1. I moved `center_char_iff` to Algebra/CharP/Subring, this is a proof for characteristic of a division ring is equal to its center's. The reason for such move is that `subring` in Algebra/CharP/Subring gives a weaker but more general statement.
2. The instance `Algebra k D` is moved to Algebra/Algebra/Basic. I don't have any convincing argument in favour of this move - I just can't find other related file that works.

acmepjz commented on Mathlib/Algebra/CharP/LinearMaps.lean:
```diff
@@ -0,0 +1,35 @@
+
+variable {R M : Type*} [CommSemiring R] [AddCommMonoid M] [Module R M]
+
+/-- For a commutative semiring `R` and a `R`-module `M`, if `M` contains an
+  element `x` that can eliminate `r` in `r • x = 0` (finding such element usually
+  depends on specific `•`), then the characteristic of `R` is equal to the
+  characteristic of `R`-linear endomorphism of `M`.-/
+instance char_eq_if {p : ℕ} [hchar : CharP R p]
+  (hreduction : ∃ x : M, ∀ r : R, r • x = 0 → r = 0) : CharP (M →ₗ[R] M) p where
```
Do you really need this instance? In my opinion this `instance` is not working with instance inferring system (only works as a `theorem`), since from the conclusion `CharP (M →ₗ[R] M) p` it cannot says `hreduction` must be satisfied.

Yu-Misaka replied:
converted to `theorem`. And do you think it is still worth keeping? The benefit of keeping it, I think, is that the injective ring homomorphism in `charP_of_injective_ringHom` might not exists for more general `R`-module `M`. So this is potentially useful somewhere else. The downside is that it kinda overlaps with `charP_of_injective_ringHom` when injective homomorphism like `Algebra.lmul` do exits, and if such function exists, `hreduction` should also be satisfied.

Yu-Misaka replied:
converted to `theorem`. Do you think it is still worth keeping?

faenuccio replied:
To add something to @acmepjz 's comment (that is perfectly appropriate): if you type
```lean
#lint
```
at the end of your file (with `char_eq_if` still marked as `instance`) you get
```
-- Found 1 error in 4 declarations (plus 62 automatically generated ones) in the current file with 15 linters

/- The `impossibleInstance` linter reports:
SOME INSTANCES HAVE ARGUMENTS THAT ARE IMPOSSIBLE TO INFER
These are arguments that are not instance-implicit and do not appear in
another instance-implicit argument or the return type. -/
#check @char_eq_if /- argument 8 hreduction : ∃ x, ∀ (r : R), r • x = 0 → r = 0 -/
```
This error tells you that type-class inference is _uncapable_ of finding `hreduction` automatically and so the system does not even accept `char_eq_if` as an `instance`. The point here is that if there are assumptions in your declaration (like `hreduction`), for it to be an `instance` it is needed that these assumptions are found _automatically_. By this, we mean that they must themselves be part of the automatic type-search tree, while if they are simply variables the system will fail to find them. The only solution would be to create a `class hreduction` whose terms are types satisfying it, and requiring at the beginning that `R` is of this class (if makes no sense here, I am just trying to explain the whole situation). As a rule-of-thumb, you can think that (apart from type variables) all variables of a class/instance must themselves be classes.

faenuccio replied:
> converted to `theorem`. Do you think it still worth keeping?

Well, why not?

Yu-Misaka replied:
> To add something to @acmepjz 's comment (that is perfectly appropriate): if you type
> 
> ```lean
> #lint
> ```
> 
> at the end of your file (with `char_eq_if` still marked as `instance`) you get
> 
> ```
> -- Found 1 error in 4 declarations (plus 62 automatically generated ones) in the current file with 15 linters
> 
> /- The `impossibleInstance` linter reports:
> SOME INSTANCES HAVE ARGUMENTS THAT ARE IMPOSSIBLE TO INFER
> These are arguments that are not instance-implicit and do not appear in
> another instance-implicit argument or the return type. -/
> #check @char_eq_if /- argument 8 hreduction : ∃ x, ∀ (r : R), r • x = 0 → r = 0 -/
> ```
> 
> This error tells you that type-class inference is _uncapable_ of finding `hreduction` automatically and so the system does not even accept `char_eq_if` as an `instance`. The point here is that if there are assumptions in your declaration (like `hreduction`), for it to be an `instance` it is needed that these assumptions are found _automatically_. By this, we mean that they must themselves be part of the automatic type-search tree, while if they are simply variables the system will fail to find them. The only solution would be to create a `class hreduction` whose terms are types satisfying it, and requiring at the beginning that `R` is of this class (if makes no sense here, I am just trying to explain the whole situation). As a rule-of-thumb, you can think that (apart from type variables) all variables of a class/instance must themselves be classes.

Thanks for the explanation! I noticed before in previous Github Action that there are lint mathlib errors but I didn't know what it was. Now it's clear, thanks again!

Yu-Misaka replied:
> > converted to `theorem`. Do you think it still worth keeping?
> 
> Well, why not?

I was thinking that, it seems possible to deduce similar result via `charP_of_injective_ringHom`. For instance, we can define map `R → (M →ₗ[R] M)` as `r ↦ (fun (x : M) ↦ r • x)`. If this map is injective, then we are done. 
And its injectivity is equivalent to `∀ x : M, ∀ r : R, r • x = 0 → r = 0`, which is slightly more strict than `hreduction : ∃ x : M, ∀ r : R, r • x = 0 → r = 0`. Plus, one might also need `R` to be a ring to prove injectivity. Hence, I think that `char_eq_if` could be useful, but I'm not sure if it is significant enough to be in mathlib.

faenuccio replied:
Well, all in all I agree that this is then probably not needed, you can simply apply `charP_of_injective_ringHom`.

acmepjz commented on Mathlib/Algebra/CharP/LinearMaps.lean:
```diff
@@ -0,0 +1,35 @@
+    refine ⟨fun h ↦ Exists.casesOn hreduction fun x hx ↦ hx (n : R) (h x),
+      fun h ↦ (congrArg (fun t ↦ ∀ (x : M), t • x = 0) h).mpr fun x ↦ zero_smul R x⟩
+
+/-- For a division ring `D` and its center `k`, `k`-linear endomorphism
+  of `D` has the same characteristic as `D`-/
+instance {D : Type*} [DivisionRing D] {p : ℕ} [hchar : CharP D p] :
+  CharP (D →ₗ[(Subring.center D)] D) p := by
+  letI : CharP (Subring.center D) p := CharP.center_char_eq_iff.1 hchar
+  refine char_eq_if (p := p) (R := (Subring.center D)) (M := D) ⟨1, ?_⟩
+  intro r hr
+  have : r • (1 : D) = r := by
+    simp [(· • ·), SMul.smul]
+  exact ZeroMemClass.coe_eq_zero.mp <| this ▸ hr
```
I think you can use [charP_of_injective_ringHom](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/CharP/Algebra.html#charP_of_injective_ringHom) (after deducing that `Subring.center D` is of characteristic p). In this way the proof becomes a one-liner.

Yu-Misaka replied:
Wow, it works, thanks!

acmepjz commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,233 @@
+  there exist a separable element xover its center-/
+theorem Jacobson_Noether (H : k ≠ (⊤ : Subring D)) :
+    ∃ x : D, x ∉ k ∧ IsSeparable k x := by
+  obtain ⟨p, hp⟩ := CharP.exists D
+  rcases CharP.char_is_prime_or_zero D p with h | h
```
I think it's better to use `ExpChar` instead, by this way the char 0 and char p proof can probably be unified.

Yu-Misaka replied:
Sorry I didn't catch it, why char 0 and char p can be unified through `ExpChar`? I tried to change the statement of `JacobsonNoether_charZero` and `JacobsonNoether_charP` into `ExpChar`. But when I attempted to use them in `Jacobson_Noether`, I still need `cases` to split the condition and apply two lemmas accordingly.

acmepjz replied:
I mean you don't need to separate the proof of `JacobsonNoether_charZero` and `JacobsonNoether_charP`. Instead, always use the proof of `JacobsonNoether_charP` but use `ExpChar` instead. The [isPurelyInseparable_iff_pow_mem](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/PurelyInseparable.html#isPurelyInseparable_iff_pow_mem) is stated using `ExpChar`, hence unifies char 0 and char p case. In the char 0 case its `ExpChar` is 1, which makes your proof of `JacobsonNoether_charP` still works for char 0 case.

Yu-Misaka replied:
> I mean you don't need to separate the proof of `JacobsonNoether_charZero` and `JacobsonNoether_charP`. Instead, always use the proof of `JacobsonNoether_charP` but use `ExpChar` instead. The [isPurelyInseparable_iff_pow_mem](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/PurelyInseparable.html#isPurelyInseparable_iff_pow_mem) is stated using `ExpChar`, hence unifies char 0 and char p case. In the char 0 case its `ExpChar` is 1, which makes your proof of `JacobsonNoether_charP` still works for char 0 case.

Thanks! I've finished major migration in [the newest commit](https://github.com/leanprover-community/mathlib4/pull/16525/commits/7555cf451433a9626b1f78617184077df5abf847).

Yu-Misaka commented:
> > > There is API for purely inseparable extension (unfortunately, only for commutative rings). Would you consider using it (after extend it to non-commutative case)?
> > 
> > 
> > I think indeed we've tried using [isPurelyInseparable_iff_pow_mem](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/PurelyInseparable.html#isPurelyInseparable_iff_pow_mem) for Fact 2 in the blog. But like you said, we need to either extend purely inseparable to non-commutative case or figure out how to define `k(\alpha)` in the notation of that blog (the difficulty here is that `IntermediateField.adjoin` demands a field over `k(\alpha)`, which isn't easy to find). This is surely an interesting idea.
> 
> The newest version of mathlib supports purely inseparable for non-commutative rings. You can try it out.

Thanks for the PR! I've already merged recent master into this branch. We'll look into it.

faenuccio commented on Mathlib/Algebra/Algebra/Basic.lean:
```diff
@@ -132,18 +132,18 @@ end CommSemiring
 section Ring
 
-variable [CommRing R]
-variable (R)
```
Can you explain this change?

acmepjz replied:
I think this is not needed, since the newly added instance has `{R : Type*}` which will override the existing `R`.

Yu-Misaka replied:
I thought `[CommRing R]` might interfere with the new instance so I moved it down. It is unnecessary if no confusion will be caused.

acmepjz commented:
> > I think this is not needed, since the newly added instance has `{R : Type*}` which will override the existing `R`.
> 
> Ah, you mean that in the new instance we do not want the `Comm` assumption on `R`?

I haven't checked the mathematics, but in the following code
```lean
variable [CommRing R]
variable (R)

...

instance {R : Type*} [Ring R] : ...
```
the code `instance {R : Type*} [Ring R]` will introduce a new `R` which is different from the existing `R` (you can check the info panel), so the original `[CommRing R]` is not carried over.

acmepjz commented on Mathlib/Algebra/Algebra/Basic.lean:
```diff
@@ -132,18 +132,18 @@ end CommSemiring
     intCast_ofNat := fun z => by simp only [Int.cast_natCast, map_natCast]
     intCast_negSucc := fun z => by simp }
 
+instance {R : Type*} [Ring R] : Algebra (Subring.center R) R :=
+  Algebra.ofModule smul_mul_assoc mul_smul_comm
+
```
I think maybe this is better?
```lean
instance {R : Type*} [Ring R] : Algebra (Subring.center R) R where
  toFun := Subtype.val
  map_one' := rfl
  map_mul' _ _ := rfl
  map_zero' := rfl
  map_add' _ _ := rfl
  commutes' := fun ⟨_, h⟩ x ↦ (Subring.mem_center_iff.1 h x).symm
  smul_def' _ _ := rfl
```
Although it's longer, but when it converts to map it is definitionally equal to `Subtype.val` and which is injective. I think this will make some of the following proofs shorter.

Yu-Misaka replied:
Indeed it might. Thanks

Yu-Misaka commented on Mathlib/Algebra/CharP/Subring.lean:
```diff
@@ -33,4 +32,11 @@ instance subring (R : Type u) [Ring R] (p : ℕ) [CharP R p] (S : Subring R) : C
+theorem center_char_eq_iff {R : Type u} [DivisionRing R] {p : ℕ} :
+    CharP R p ↔ CharP (Subring.center R) p :=
+  (@RingHom.charP_iff (Subring.center R) R _ _ (algebraMap (Subring.center R) R)
+    (NoZeroSMulDivisors.algebraMap_injective (Subring.center R) R) p).symm
```
implemented, thanks!

Yu-Misaka commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,233 @@
+* <https://ysharifi.wordpress.com/2011/09/30/the-jacobson-noether-theorem/>
+-/
+
+-- *Filippo* This should probably be moved to another file.
```
moved to Algebra/Field/Power.lean as there are relevant results concerning division ring.

Yu-Misaka reviewed (commented):
Greetings! In this commit, I added section `purelyInsep`, in which I tried to switch to `ExpChar` and purely inseparable. By such, `JacobsonNoether_charZero` and `JacobsonNoether_charP` are unified. 
However, I wasn't able to fill in the two `sorry` in `temp_exists_pow_mem_center_ofInseparable`, line 209. These two `sorry` are used to show that `Algebra.adjoin k {a}` (`k(a)` in the notation of the blog) is purely separable over `k`. The difficulties here are to prove `k(a)` is integral over `k`, and resolve the image of `algebraMap`. Plus, I noticed that the new instance in LinearMaps seems to be slow.

acmepjz commented:
> I wasn't able to fill in the two `sorry` in `temp_exists_pow_mem_center_ofInseparable`, line 209. These two `sorry` are used to show that `Algebra.adjoin k {a}` (`k(a)` in the notation of the blog) is purely separable over `k`. The difficulties here are to prove `k(a)` is integral over `k`, and resolve the image of `algebraMap`.

I don't think you need to tackle `Algebra.adjoin` things here, since they are already solved in separable/purely inseparable APIs. You may start with the following code:
```lean
import Mathlib.FieldTheory.PurelyInseparable

open LinearMap in
theorem test1
    {k D : Type*} [Field k] [DivisionRing D] [Algebra k D] [Algebra.IsAlgebraic k D]
    (hcenter : Subalgebra.center k D = ⊥)
    (hneq : (⊥ : Subalgebra k D) ≠ ⊤) :
    ∃ x : D, x ∉ (⊥ : Subalgebra k D) ∧ IsSeparable k x := by
  classical
  obtain ⟨q, _⟩ := ExpChar.exists k
  by_contra! H
  have h1 : IsPurelyInseparable k D :=
    ⟨Algebra.IsAlgebraic.isIntegral, fun x ↦ not_imp_not.1 (H x)⟩
  rw [isPurelyInseparable_iff_pow_mem k q] at h1
  sorry
```

Yu-Misaka commented:
> > I wasn't able to fill in the two `sorry` in `temp_exists_pow_mem_center_ofInseparable`, line 209. These two `sorry` are used to show that `Algebra.adjoin k {a}` (`k(a)` in the notation of the blog) is purely separable over `k`. The difficulties here are to prove `k(a)` is integral over `k`, and resolve the image of `algebraMap`.
> 
> I don't think you need to tackle `Algebra.adjoin` things here, since they are already solved in separable/purely inseparable APIs. You may start with the following code:
> 
> ```lean
> import Mathlib.FieldTheory.PurelyInseparable
> 
> open LinearMap in
> theorem test1
>     {k D : Type*} [Field k] [DivisionRing D] [Algebra k D] [Algebra.IsAlgebraic k D]
>     (hcenter : Subalgebra.center k D = ⊥)
>     (hneq : (⊥ : Subalgebra k D) ≠ ⊤) :
>     ∃ x : D, x ∉ (⊥ : Subalgebra k D) ∧ IsSeparable k x := by
>   classical
>   obtain ⟨q, _⟩ := ExpChar.exists k
>   by_contra! H
>   have h1 : IsPurelyInseparable k D :=
>     ⟨Algebra.IsAlgebraic.isIntegral, fun x ↦ not_imp_not.1 (H x)⟩
>   rw [isPurelyInseparable_iff_pow_mem k q] at h1
>   sorry
> ```
> 
> First you can conclude `IsPurelyInseparable k D` just from the definition of purely inseparable in mathlib. Then you use `isPurelyInseparable_iff_pow_mem` to conclude that every element of `D` can be powered into `k`.

Indeed, I can bypass `Algebra.adjoin`, thanks! I've finished migration of the first statement, and will try out the second statement.

Yu-Misaka commented:
I added in namespace `canary` some trials regarding translation between two statements inspired by @faenuccio. Now the only thing remains is proving `IsSeparable L x` from `IsSeparable (center D) x` given `L ≃+* (center D)`. I wonder if we have any method in Lean to "view isomorphic structures as the same thing" as we sometimes do in math.

acmepjz commented:
> Now the only thing remains is proving `IsSeparable L x` from `IsSeparable (center D) x` given `L ≃+* (center D)`. I wonder if we have any method in Lean to "view isomorphic structures as the same thing" as we sometimes do in math.

Lean is very explicit on such things (compared to human doing math, which will occasionally cause subtle errors). Maybe you can try [IsSeparable.of_equiv_equiv](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Separable.html#IsSeparable.of_equiv_equiv).

Yu-Misaka commented:
> > Now the only thing remains is proving `IsSeparable L x` from `IsSeparable (center D) x` given `L ≃+* (center D)`. I wonder if we have any method in Lean to "view isomorphic structures as the same thing" as we sometimes do in math.
> 
> Lean is very explicit on such things (compared to human doing math, which will occasionally cause subtle errors). Maybe you can try [IsSeparable.of_equiv_equiv](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Separable.html#IsSeparable.of_equiv_equiv).
> 
> Oh, it requires fields... maybe we should generalize that result first...

Good idea! I found that at least in our application here, we can generalize [IsSeparable.of_equiv_equiv](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Separable.html#IsSeparable.of_equiv_equiv) via generalizing [IsSeparable.tower_top](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Separable.html#IsSeparable.tower_top), which can be done by removing commutativity in [minpoly.dvd_map_of_isScalarTower](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Minpoly/Field.html#minpoly.dvd_map_of_isScalarTower). I've wrote a sketch `IsSeparable.of_equiv_equiv'` in namespace `canary`.

acmepjz commented:
> we can generalize [IsSeparable.of_equiv_equiv](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Separable.html#IsSeparable.of_equiv_equiv) via generalizing [IsSeparable.tower_top](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Separable.html#IsSeparable.tower_top), which can be done by removing commutativity in [minpoly.dvd_map_of_isScalarTower](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Minpoly/Field.html#minpoly.dvd_map_of_isScalarTower).

You are welcomed to submit a separate PR for generalizing these.

Yu-Misaka commented:
Thanks! Though I still have some questions before that. I've moved related codes to namespace `separable_canary`. The first question is that I want to have an instance `Algebra A₂ B₁` in line 200, but if I switch `def` to `instance`, it immediately yields error `cannot find synthesization order for instance @instAlgebra`. The second is that I want `IsSeparable.of_equiv_equiv'` and `IsSeparable.of_equiv_equiv''` to use variables defined at the beginning of the namespace, but once I omit them in the statement, the infer system seems to be confused.

acmepjz commented:
If it's me, I'll write the instance `Algebra A₂ B₁` into the proof of `IsSeparable.of_equiv_equiv` instead of outside (because it only uses inside this proof, put it outside will confuse typeclass inferencing system). Like this:
```lean
lemma IsSeparable.of_equiv_equiv₀ ... ... :=
  ......
  letI : Algebra A₂ B₁ := {
    ......
  }
  ...
```

Yu-Misaka commented:
> If it's me, I'll write the instance `Algebra A₂ B₁` into the proof of `IsSeparable.of_equiv_equiv` instead of outside (because it only uses inside this proof, put it outside will confuse typeclass inferencing system). Like this:
> 
> ```lean
> lemma IsSeparable.of_equiv_equiv₀ ... ... :=
>   ......
>   letI : Algebra A₂ B₁ := {
>     ......
>   }
>   ...
> ```
> 
> Maybe you can check the original proof
> 
> https://github.com/leanprover-community/mathlib4/blob/4e3fd1ae9518ed2894bfc00592c81a3ee2b6180e/Mathlib/FieldTheory/Separable.lean#L687-L697
> the line 689 defines the instance `Algebra A₂ B₁`.

Indeed that could help. I created a PR on it so we might discuss more about it there. The problem however, is that I might use this instance in the statement. For example in `IsSeparable.of_equiv'` (line 725) of the new PR, I need to state `IsSeparable A₂ x` - this requires `Algebra A₂ B₁`. And the original proof could fail here since `.toAlgebra` requires commutative `B₁`.

Yu-Misaka commented:
Greetings! Sorry I have been stuck here for a while and I didn't manage to advance. The problem now is to prove ```(algebraMap L D).comp ↑equiv.symm = algebraMap { x // x ∈ center D } D```, where ```equiv``` is equivalence ```L ≃+* { x // x ∈ center D }```. Can any one help? Thanks a lot in advance!

faenuccio commented:
@acmepjz I think this might be a good moment to share the proof you showed me in the DM of Sep14th of your `test1` to prove the `example` on ll. 198 of `Mathlib\FieldTheory\JacobsonNoether.lean`, no? I can do this, but since this is your code, I'd rather leave it to you 😄

acmepjz commented:
> I think this might be a good moment to share the proof you showed me in the DM of Sep14th of your `test1` to prove the `example` on ll. 198 of `Mathlib\FieldTheory\JacobsonNoether.lean`, no? I can do this, but since this is your code, I'd rather leave it to you 😄

Sure, but my proof is in the other direction
```lean
import Mathlib.FieldTheory.PurelyInseparable

open LinearMap in
theorem test1
    {k D : Type*} [Field k] [DivisionRing D] [Algebra k D] [Algebra.IsAlgebraic k D]
    (hcenter : Subalgebra.center k D = ⊥)
    (hneq : (⊥ : Subalgebra k D) ≠ ⊤) :
    ∃ x : D, x ∉ (⊥ : Subalgebra k D) ∧ IsSeparable k x := by
  -- omitted for now
  sorry

instance {R : Type*} [Ring R] : Algebra (Subring.center R) R where
  toFun := Subtype.val
  map_one' := rfl
  map_mul' _ _ := rfl
  map_zero' := rfl
  map_add' _ _ := rfl
  commutes' r x := (Subring.mem_center_iff.1 r.2 x).symm
  smul_def' _ _ := rfl

theorem test2 {D : Type*} [DivisionRing D]
    [Algebra.IsAlgebraic (Subring.center D) D]
    (hneq : Subring.center D ≠ ⊤) :
    ∃ x : D, x ∉ Subring.center D ∧ IsSeparable (Subring.center D) x := by
  have := test1 (k := Subring.center D) (D := D) ?_ ?_
  · obtain ⟨x, hx, hsep⟩ := this
    exact ⟨x, by simpa using (show x ∉ Set.range Subtype.val from hx), hsep⟩
  · ext x
    change x ∈ Subring.center D ↔ x ∈ Set.range Subtype.val
    simp
  · contrapose! hneq
    apply SetLike.coe_injective
    simpa using (show Set.range Subtype.val = ⊤ from congr(SetLike.coe $(hneq)))
```
But I think the author asks for `test2 -> test1`, which I'm not think of yet.

Note that my proof does not use `IsSeparable.of_equiv_equiv`.

> in fact I don't quite understand why we need the diagram to be commutative in `IsSeparable.of_equiv_equiv`

Maybe you can check its proof to see what will happen without the condition that the diagram is commutative.

acmepjz commented:
> The problem now is to prove `(algebraMap L D).comp ↑equiv.symm = algebraMap { x // x ∈ center D } D`, where `equiv` is equivalence `L ≃+* { x // x ∈ center D }`.

Can you try `by ext; rfl` or `by ext; simp`?

acmepjz commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,327 @@
+  set ψ : L ≃+* (⊥ : Subalgebra L D) := ⟨((Algebra.botEquiv L D).toRingEquiv).symm,
+      fun _ _ ↦ by simp only [Equiv.toFun_as_coe, EquivLike.coe_coe, map_mul],
+      fun _ _ ↦ by simp only [Equiv.toFun_as_coe, EquivLike.coe_coe, map_add]⟩
+  set φ : (⊥ : Subalgebra L D) ≃+* (center D) := hcenter ▸ RingEquiv.refl (Subalgebra.center L D)
```
It's best to use `Subalgebra.equivOfEq` here. The `▸` probably will erase the actual definition of `φ`.

Yu-Misaka replied:
Indeed that make it easier to reach the definition, thanks!

faenuccio commented:
> > The problem now is to prove `(algebraMap L D).comp ↑equiv.symm = algebraMap { x // x ∈ center D } D`, where `equiv` is equivalence `L ≃+* { x // x ∈ center D }`.
> 
> Can you try `by ext; rfl` or `by ext; simp`?

I proved both, and I also tried to expand a bit the definitions, with no result. Probably the suggestion of using the `Subalgebra.equivOfEq` is the right one, haven't checked it.

faenuccio commented:
> > I think this might be a good moment to share the proof you showed me in the DM of Sep14th of your `test1` to prove the `example` on ll. 198 of `Mathlib\FieldTheory\JacobsonNoether.lean`, no? I can do this, but since this is your code, I'd rather leave it to you 😄
> 
> Sure, but my proof is in the other direction
> 
> ```lean
> import Mathlib.FieldTheory.PurelyInseparable
> 
> open LinearMap in
> theorem test1
>     {k D : Type*} [Field k] [DivisionRing D] [Algebra k D] [Algebra.IsAlgebraic k D]
>     (hcenter : Subalgebra.center k D = ⊥)
>     (hneq : (⊥ : Subalgebra k D) ≠ ⊤) :
>     ∃ x : D, x ∉ (⊥ : Subalgebra k D) ∧ IsSeparable k x := by
>   -- omitted for now
>   sorry
> 
> instance {R : Type*} [Ring R] : Algebra (Subring.center R) R where
>   toFun := Subtype.val
>   map_one' := rfl
>   map_mul' _ _ := rfl
>   map_zero' := rfl
>   map_add' _ _ := rfl
>   commutes' r x := (Subring.mem_center_iff.1 r.2 x).symm
>   smul_def' _ _ := rfl
> 
> theorem test2 {D : Type*} [DivisionRing D]
>     [Algebra.IsAlgebraic (Subring.center D) D]
>     (hneq : Subring.center D ≠ ⊤) :
>     ∃ x : D, x ∉ Subring.center D ∧ IsSeparable (Subring.center D) x := by
>   have := test1 (k := Subring.center D) (D := D) ?_ ?_
>   · obtain ⟨x, hx, hsep⟩ := this
>     exact ⟨x, by simpa using (show x ∉ Set.range Subtype.val from hx), hsep⟩
>   · ext x
>     change x ∈ Subring.center D ↔ x ∈ Set.range Subtype.val
>     simp
>   · contrapose! hneq
>     apply SetLike.coe_injective
>     simpa using (show Set.range Subtype.val = ⊤ from congr(SetLike.coe $(hneq)))
> ```
> 
> But I think the author asks for `test2 -> test1`, which I'm not think of yet.

Ah right! Well, they might be happy to see a proof of `test1` directly and actually all of the code you gave me, by now...😄

Yu-Misaka commented:
Thanks all for your kind help @faenuccio @acmepjz! I sincerely appreciate that. Also, @FR-vdash-bot managed to close the second statement in a rather neat way as the above commit suggests.

faenuccio commented on Mathlib/Algebra/CharP/LinearMaps.lean:
```diff
@@ -0,0 +1,65 @@
+  element `x` that can eliminate `r` in `r • x = 0` (finding such element usually
+  depends on specific `•`), then the characteristic of `R` is equal to the
+  characteristic of `R`-linear endomorphism of `M`.-/
+theorem char_eq_if {p : ℕ} [hchar : CharP R p]
```
See above my comment about this name.

faenuccio commented on Mathlib/Algebra/CharP/LinearMaps.lean:
```diff
@@ -0,0 +1,65 @@
+    refine ⟨fun h ↦ Exists.casesOn hreduction fun x hx ↦ hx n (h x),
+      fun h ↦ (congrArg (fun t ↦ ∀ x, t • x = 0) h).mpr fun x ↦ zero_smul R x⟩
+
+/-- For a division ring `D` and its center `k`, `k`-linear endomorphism
```
```suggestion
/-- For a division ring `D` and its center `k`, the ring of `k`-linear endomorphisms
```

faenuccio commented on Mathlib/Algebra/CharP/LinearMaps.lean:
```diff
@@ -0,0 +1,65 @@
+  cast_eq_zero_iff' := by
+    intro n
+    replace hchar := hchar.1 n
+    have reduce : (n : M →ₗ[R] M) = (n : R) • 1 := by
+      simp only [Nat.cast_smul_eq_nsmul, nsmul_eq_mul, mul_one]
```
```suggestion
    have exact : (n : M →ₗ[R] M) = (n : R) • 1 := by
      simp only [Nat.cast_smul_eq_nsmul, nsmul_eq_mul, mul_one]
```

faenuccio commented on Mathlib/Algebra/CharP/LinearMaps.lean:
```diff
@@ -0,0 +1,65 @@
+    (hreduction : ∃ x : M, ∀ r : R, r • x = 0 → r = 0) : CharP (M →ₗ[R] M) p where
+  cast_eq_zero_iff' := by
+    intro n
+    replace hchar := hchar.1 n
```
This line is actually useless, you can simply replace `hchar.symm` by `← hchar.1` two lines below.

faenuccio commented on Mathlib/Algebra/Field/Power.lean:
```diff
@@ -17,6 +17,12 @@ beyond those needed for the basic definition.
 variable {α : Type*}
 
+/-- a⁻¹ * d ^ s * a = (a⁻¹ * d * a) ^ s when a is not zero -/
+lemma conj_nonComm_Algebra [DivisionSemiring α] (s : ℕ) (a d : α) (ha : a ≠ 0) :
```
Can you try making both `s` and `a,d` implicit (so in curly brackets `{` and `}` rather than in usual ones `(` and `)`)?

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,209 @@
+
+This file contains a proof of the Jacobson-Noether theorem and some auxiliary lemmas.
+Here we discuss different cases of characteristics of
+the noncommutative division algebra `D` with the center `k`.
```
```suggestion
the noncommutative division algebra `D` with center `k`.
```

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,209 @@
+
+## Main Results
+
+- `Jacobson_Noether` : Let `D` be a noncommutative division algebra with the center `k`.
```
```suggestion
- `Jacobson_Noether` : Let `D` be a noncommutative division algebra with center `k`.
```

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,209 @@
+variable {D : Type*} [DivisionRing D]
+
+local notation3 "k" => (Subring.center D)
+local notation3 "f" => (LinearMap.mulLeft k)
```
I would suggest that you remove `f` and `g` altogether. They already have a name in mathlib and if you `open LinearMap` this name is not very long), and their commutativity is also in the library, so you should rather use it directly rather than creating a new, ad-hoc lemma. You can keep `δ` and the related business, nevertheless (marking them `private` is certainly a good idea).

Yu-Misaka commented:
@faenuccio Greetings! Thanks, I've implemented all suggestions
(btw would you please help review #17147? thanks!

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,206 @@
+  have inter : (δ a) ^ (p ^ m) = 0 := by
+    refine LinearMap.ext_iff.2 ?_
+    intro x
+    rw [δ_def' a, sub_pow_expChar_pow_of_commute p m (commute_mulLeft_right a a)]
```
```suggestion
    rw [δ, Pi.sub_apply, sub_pow_expChar_pow_of_commute p m (commute_mulLeft_right a a)]
```

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,206 @@
+
+variable (D) in
+/-- For a non-commutative finite dimensional division algebra D (with base ring being its center),
+  there exist a separable element xover its center-/
```
```suggestion
  there exist an element `x` that is separable over its center-/
```

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,206 @@
+
+private lemma δ_def (a x : D) : δ a x = mulLeft k a x - mulRight k a x := rfl
+
+private lemma δ_def' (a : D) : δ a = mulLeft k a - mulRight k a := rfl
```
```suggestion
```

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,206 @@
+  | succ _ h => rw [Function.iterate_succ', Function.comp_apply, h,
+    pow_succ', mul_assoc, mulLeft_apply]
+
+private lemma r_pow (a : D) (n : ℕ) : ∀ x : D, ((mulRight k a) ^ n).1 x = x * (a ^ n) := by
```
```suggestion
private lemma r_pow (a : D) (n : ℕ) : ∀ x : D, ((mulRight k a) ^ n) x = x * (a ^ n) := by
```
Type class inference is able to see through the projection from this ring hom down to a `AddHom`, so please remove this `.1`; you can then remove the `coe_toAddHom` from the `rw`, and do the same for `r_pow`. In the applications around line 112, it suffices again to remove the projection `.1`.

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,206 @@
+lemma exists_pow_mem_center_ofInseparable (p : ℕ) [hchar : ExpChar D p] (a : D)
+    (hinsep : ∀ x : D, IsSeparable k x → x ∈ k) : ∃ n, a ^ (p ^ n) ∈ k := by
+  have := (@isPurelyInseparable_iff_pow_mem k D _ _ _ _ p (ExpChar.center_expChar_iff.1 hchar)).1
+  have bridge : ∀ x : k, (algebraMap k D) x = x := fun _ ↦ rfl
```
This `bridge` seems useless.

Yu-Misaka replied:
Indeed `bridge` and `bridge'` can both be omitted, thanks!

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,206 @@
+lemma exists_pow_mem_center_ofInseparable (p : ℕ) [hchar : ExpChar D p] (a : D)
+    (hinsep : ∀ x : D, IsSeparable k x → x ∈ k) : ∃ n, a ^ (p ^ n) ∈ k := by
+  have := (@isPurelyInseparable_iff_pow_mem k D _ _ _ _ p (ExpChar.center_expChar_iff.1 hchar)).1
+  have bridge : ∀ x : k, (algebraMap k D) x = x := fun _ ↦ rfl
```
you will only need this once; so instead of adding it here, you can use as a `rw` at `pure` later on (or use the `▸`)

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,206 @@
+  exact ⟨n, Set.mem_of_subset_of_mem this <| Set.mem_range.2 ⟨m, hm⟩⟩
+
+/-- If `D` is a purely inseparable extension of `k` with characteristic `p`,
+  then for every element `a` of `D\k`, there exists a natural number `n`
```
```suggestion
  then for every element `a` of `D \ k`, there exists a natural number `n`
```

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,206 @@
+/-- If `D` is a purely inseparable extension of `k` with characteristic `p`,
+  then for every element `a` of `D`, there exists a natural number `n`
+  such that `a ^ (p ^ n)` is contained in `k`. -/
+lemma exists_pow_mem_center_ofInseparable (p : ℕ) [hchar : ExpChar D p] (a : D)
```
```suggestion
lemma exists_pow_mem_center_of_inseparable (p : ℕ) [hchar : ExpChar D p] (a : D)
```

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,206 @@
+/-- If `D` is a purely inseparable extension of `k` with characteristic `p`,
+  then for every element `a` of `D\k`, there exists a natural number `n`
+  **greater than 0** such that `a ^ (p ^ n)` is contained in `k`. -/
+lemma exists_pow_mem_center_ofInseparable' (p : ℕ) [ExpChar D p] {a : D}
```
```suggestion
lemma exists_pow_mem_center_of_inseparable' (p : ℕ) [ExpChar D p] {a : D}
```

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,206 @@
+  obtain ⟨n, hn⟩ := exists_pow_mem_center_ofInseparable p a hinsep
+  by_cases nzero : n = 0
+  · rw [nzero, pow_zero, pow_one] at hn
+    exact False.elim <| ha hn
```
```suggestion
    exact (ha hn).elim
```

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,206 @@
+    exact False.elim <| ha hn
+  · exact ⟨n, ⟨by omega, hn⟩⟩
+
+/-- If `D` is a purely inseparable extension of `k` with characteristic `p`,
```
```suggestion
/-- If `D` is a purely inseparable extension of `k` of characteristic `p`,
```

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,206 @@
+  · exact ⟨n, ⟨by omega, hn⟩⟩
+
+/-- If `D` is a purely inseparable extension of `k` with characteristic `p`,
+  then for every element `a` of `D\k`, there exists a natural number `m`
```
```suggestion
  then for every element `a` of `D \ k`, there exists a natural number `m`
```

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,206 @@
+
+/-- If `D` is a purely inseparable extension of `k` with characteristic `p`,
+  then for every element `a` of `D\k`, there exists a natural number `m`
+  greater than 0 such that linear map `(a * x - x * a) ^ n = 0` for
```
```suggestion
  greater than 0 such that `(a * x - x * a) ^ n = 0` (as linear maps) for
```

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,206 @@
+/-- If `D` is a purely inseparable extension of `k` with characteristic `p`,
+  then for every element `a` of `D\k`, there exists a natural number `m`
+  greater than 0 such that linear map `(a * x - x * a) ^ n = 0` for
+  any `n` greater than `(p ^ n)`. -/
```
```suggestion
  every `n` greater than `(p ^ m)`. -/
```

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,206 @@
+  refine ⟨m, ⟨hm.1, ?_⟩⟩
+  intro n hn
+  have inter : (δ a) ^ (p ^ m) = 0 := by
+    refine LinearMap.ext_iff.2 ?_
+    intro x
```
```suggestion
    ext x
```

faenuccio reviewed (commented):
OK, I have added more comments and stopped short of the two proofs of JN. I will also have a look at #17147 once this PR will get merged (but please re-ping me if I forget).

Yu-Misaka commented on Mathlib/Algebra/CharP/LinearMaps.lean:
```diff
@@ -0,0 +1,60 @@
+# Characteristics of Linear Maps
+
+This file contains properties of the characteristic of the ring of linear maps.
+Characteristics of linear maps are determined by its scalar.
```
I meant to say that for an `R`-module `M`, the characteristic of the ring of `R`-linear endomorphisms of `M` is determined by `R`, which is the "scalar" of the scalar multiplication of `M`...
But in the hindsight I think this sentence looks odd and useless, maybe we can remove it?

faenuccio replied:
Or you can replace "scalar"  by "base ring".

Yu-Misaka replied:
Good idea! implemented.

Yu-Misaka commented on Mathlib/Algebra/CharP/LinearMaps.lean:
```diff
@@ -0,0 +1,60 @@
+
+- `CharP_if` : For a commutative semiring `R` and a `R`-module `M`,
+  the characteristic of `R` is equal to the characteristic of `R`-linear endomorphism of `M` when
+  `M` contains an element `x` that can eliminate `r` in `r • x = 0`
```
Sure, I'll explain more details.

Yu-Misaka commented on Mathlib/Algebra/Field/Power.lean:
```diff
@@ -17,6 +17,12 @@ beyond those needed for the basic definition.
 
+/-- a⁻¹ * d ^ s * a = (a⁻¹ * d * a) ^ s when a is not zero -/
+lemma conj_nonComm_Algebra [DivisionSemiring α] {s : ℕ} {a d : α} (ha : a ≠ 0) :
+    a⁻¹ * d ^ s * a = (a⁻¹ * d * a) ^ s := by
```
Sorry I didn't catch up here, wouldn't `a = 0` cause trouble when `s` is also 0? In which case, the left hand side is 0 while the right hand side is `0 ^ 0 = 1`. (btw I just noticed that mathlib set the inverse of `0` as `0` for division ring, is it a convention?)

faenuccio replied:
Oh right, thanks! Concerning `0 ^ 0` you're perfectly right, due to [zero_pow_eq_one₀](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/GroupWithZero/Basic.html#zero_pow_eq_one%E2%82%80) so please forget my comment!

For the second, I think that you can indeed call it `DivisionSemiring.conj_pow`, perhaps putting everything in the `namespace DivisionSemiring` _provided it already exists._

erdOne replied:
This works for all groups with zeroes right? and it should be called `conj_pow₀` in that case.

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,198 @@
+    refine ⟨Nat.find exist, ⟨(Nat.find_spec exist).1, ?_, (Nat.find_spec exist).2⟩⟩
+    set t := (Nat.find exist - 1 : ℕ) with ht
+    by_cases choice : 0 < t
+    · have := @Nat.find_min (H := exist) _ _ t ?_
+      · exact (@Nat.sub_add_cancel (Nat.find exist) 1 (by omega) ▸ ht ▸ not_and.1 this) choice
+      · exact Nat.sub_one_lt <| ne_of_gt (Nat.find_spec exist).1
```
```suggestion
    · convert (ne_eq _ _) ▸ not_and.mp (@Nat.find_min (H := exist) _ _ t (by omega)) choice
      omega
```

faenuccio reviewed (commented):
Boarding my flight now, but I'll continue the review later

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,198 @@
+    · have := @Nat.find_min (H := exist) _ _ t ?_
+      · exact (@Nat.sub_add_cancel (Nat.find exist) 1 (by omega) ▸ ht ▸ not_and.1 this) choice
+      · exact Nat.sub_one_lt <| ne_of_gt (Nat.find_spec exist).1
+    · rw [not_lt, Nat.le_zero] at choice
+      have := Nat.eq_add_of_sub_eq (Nat.find_spec exist).1 ht.symm
+      simp only [gt_iff_lt, choice, Nat.succ_eq_add_one, zero_add] at this
+      rw [this, pow_one]
+      exact hb1
```
```suggestion
    · suffices h_find: Nat.find exist = 1 by
        rwa [h_find, pow_one]
      rw [not_lt, Nat.le_zero, ht, Nat.sub_eq_zero_iff_le] at choice
      linarith [(Nat.find_spec exist).1]
```

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,198 @@
+  letI : Invertible c := ⟨c⁻¹, inv_mul_cancel₀ (hb.1), mul_inv_cancel₀ (hb.1)⟩
+  -- We prove that `c` is commute with `a`.
+  have hc : c * a = a * c := by
+    symm; apply eq_of_sub_eq_zero
```
This `symm` shows that you set up your equations in the wrong order (both `hc` and `hb.2`): better change them and avoid this `symm` here.

Yu-Misaka replied:
(Sorry for the delay, was dealing with some school stuff these two days)
Indeed, I've reverted ```hc```. Regarding ```hb.2```, should we change that too? Given that both left and right parts of ```hb``` are used.

faenuccio reviewed (commented):
I have posted more comments for the first proof of JN: please implement them and then I'll resume the review.

Yu-Misaka commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,198 @@
+    classical
+    refine ⟨Nat.find exist, ⟨(Nat.find_spec exist).1, ?_, (Nat.find_spec exist).2⟩⟩
+    set t := (Nat.find exist - 1 : ℕ) with ht
+    by_cases choice : 0 < t
```
True, ```choice``` is somewhat confusing...

Yu-Misaka commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,198 @@
+    show a * ha.choose - ha.choose * a ≠ 0
+    simpa only [ne_eq, sub_eq_zero] using Ne.symm ha.choose_spec
+  -- We find a maximum natural number `n` such that `(δ a) ^ n b ≠ 0`.
+  obtain ⟨n, hn, hb⟩ : ∃ n > 0, ((δ a) ^ n) b ≠ 0 ∧ ((δ a) ^ (n + 1)) b = 0 := by
```
It seems Lean can't handle ```∃ 0 < n``` and it yields error ```elaboration function for 'Lean.«termSatisfies_binder_pred%__»' has not been implemented```.
Is it ok to wirte ```∃ n, (0 < n) ∧ ((δ a) ^ n) b ≠ 0 ∧ ((δ a) ^ (n + 1)) b = 0``` instead?

faenuccio replied:
Sure!

Yu-Misaka commented on Mathlib/Algebra/CharP/LinearMaps.lean:
```diff
@@ -0,0 +1,60 @@
+
+variable {R M : Type*} [CommSemiring R] [AddCommMonoid M] [Module R M]
+
+/-- For a commutative semiring `R` and a `R`-module `M`, if `M` contains an
```
oops, sorry I just noticed that "eliminate..." appeared twice, I've changed it to ```that can deduce `r = 0` given `r • x = 0` ``` as in the above doc.

Yu-Misaka commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,198 @@
+variable (D) in
+/-- For a non-commutative finite dimensional division algebra D (with base ring being its center),
+  there exist an element `x` that is separable over its center-/
+theorem Jacobson_Noether (H : k ≠ (⊤ : Subring D)) :
```
Sure, that would definitely be more explict.

By the way, apologize if I seems silly but why is the annotation talking about ```finite dimensional division algebra D``` instead of something like ```algebraic division algebra D```? Is finite equivalent to algebraic in this case?

Yu-Misaka commented:
@faenuccio Greetings! Thanks for your review, and I've implemented most of the comments.
For `∃ n > 0, ...`, I changed it to `∃ n, (0 < n) ∧ ...` since `∃ 0 < n` doesn't seem to work. I also left `hb` unchanged because both parts of it are used. In addition, I changed the "finite dimensional" in JN's annotation to "algebraic" since they might not be equivalent in my view. But please tell me if I'm wrong or I mistook anything.

faenuccio commented on Mathlib/Algebra/CharP/LinearMaps.lean:
```diff
@@ -0,0 +1,57 @@
+variable {R M : Type*} [CommSemiring R] [AddCommMonoid M] [Module R M]
+
+/-- For a commutative semiring `R` and a `R`-module `M`, if `M` contains an
+  element `x` that can deduce `r = 0` given `r • x = 0` (finding such element usually
```
```suggestion
  element `x` such that `r • x = 0` implies `r = 0` (finding such element usually
```

Yu-Misaka commented:
The compilation works fine, but there seems to be some git errors. I'll try merge master

faenuccio commented:
> @faenuccio Greetings! Thanks for your review, and I've implemented most of the comments. For `∃ n > 0, ...`, I changed it to `∃ n, (0 < n) ∧ ...` since `∃ 0 < n` doesn't seem to work. I also left `hb` unchanged because both parts of it are used. In addition, I changed the "finite dimensional" in JN's annotation to "algebraic" since they might not be equivalent in my view. But please tell me if I'm wrong or I mistook anything.

I don't see this change, can you point out to the line where you implemented it? At any rate, every division algebra that is algebraic over its center is f.d. (see T. J. Laflίey, Commutative subalgebras of infinite dimensional algebras, Bull. London Math. Soc, 5 (1973), 312-314.) but this is not implemented in mathlb, asaik.

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,215 @@
+  rw [(Nat.sub_eq_iff_eq_add hn).1 rfl, pow_add, inter, mul_zero]
+
+variable (D) in
+/-- Jacobson-Noether theorem: For a non-commutative algebraic
+  division algebra `D` (with base ring being its center `k`), then
+  there exists an element `x` of `D \ k` that is separable over `k`. -/
```
```suggestion
/-- Jacobson-Noether theorem: For a non-commutative division algebra 
  `D` that is algebraic over its center `k`, there exists an element
  `x` of `D \ k` that is separable over `k`. -/
```

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,215 @@
+    nth_rw 3 [← mul_one a]
+    rw [hd_def, ← eq1, mul_sub, mul_assoc _ _ a, sub_right_inj, hc',
+      ← mul_assoc, ← mul_assoc, ← mul_assoc]
+  -- This then derives a contradiction.
```
```suggestion
  -- This then yields a contradiction.
```

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,215 @@
+  exact one_ne_zero deq
+
+open Subring Algebra in
+/-- Jacobson-Noether theorem: For a non-commutative algebraic
+  division algebra `D` (with base ring being a field `L`), if the center of
+  `D` over `L` is `L`, then there exist an element `x` of `D \ L`
+  that is separable over `L`. -/
```
```suggestion
/-- Jacobson-Noether theorem: For a non-commutative division algebra `D` 
  that is algebraic over a field `L`, if the center of
  `D` coincides with `L`, then there exist an element `x` of `D \ L`
  that is separable over `L`. -/
```

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,215 @@
+    (hcenter : Subalgebra.center L D = ⊥) (hneq : (⊥ : Subalgebra L D) ≠ ⊤) :
+    ∃ x : D, x ∉ (⊥ : Subalgebra L D) ∧ IsSeparable L x := by
+  have ntrivial : center D ≠ ⊤ :=
+    congr(Subalgebra.toSubring $hcenter).trans_ne (Subalgebra.toSubring_injective.ne hneq)
```
Can you explain (to me) this syntax with `congr`, no space before the `(` and the `$`?

Yu-Misaka replied:
This part was done by @FR-vdash-bot as I mentioned in commit [bb0c028](https://github.com/leanprover-community/mathlib4/pull/16525/commits/bb0c02893475e03e46ef653f6886b90cc34cf7a9). So I am new to this syntax😂

From my understanding, it acts `Subalgebra.toSubring` on both parts of the equality `hcenter`. Thus `congr(Subalgebra.toSubring $hcenter)` yields `(Subalgebra.center L D).toSubring = ⊥.toSubring`, kinda like `apply_fun` in equality.

So here we can also close the target via
```
  have ntrivial : center D ≠ ⊤ := by
    apply_fun Subalgebra.toSubring at hcenter
    exact hcenter.trans_ne (Subalgebra.toSubring_injective.ne hneq)
```
or
```
  have ntrivial : center D ≠ ⊤ := by
    apply congrArg Subalgebra.toSubring at hcenter
    exact hcenter.trans_ne (Subalgebra.toSubring_injective.ne hneq)
```
Though I don't quite understand the 
>This is able to make use of the expected type, for example `(congr(_ + $h) : 1 + _ = _)` with `h : x = y` gives `1 + x = 1 + y`.


in its documentation.

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,215 @@
+    congr(Subalgebra.toSubring $hcenter).trans_ne (Subalgebra.toSubring_injective.ne hneq)
+  set φ := Subalgebra.equivOfEq (⊥ : Subalgebra L D) (Subalgebra.center L D) hcenter.symm
+  set equiv : L ≃+* (center D) := ((botEquiv L D).symm.trans φ).toRingEquiv
+  letI : Algebra L (center D) := equiv.toRingHom.toAlgebra
+  letI : Algebra (center D) L := equiv.symm.toRingHom.toAlgebra
+  haveI : IsScalarTower L (center D) D := .of_algebraMap_eq fun _ ↦ rfl
+  haveI : IsScalarTower (center D) L D := .of_algebraMap_eq fun x ↦ by
```
```suggestion
  let _ : Algebra L (center D) := equiv.toRingHom.toAlgebra
  let _ : Algebra (center D) L := equiv.symm.toRingHom.toAlgebra
  have _ : IsScalarTower L (center D) D := .of_algebraMap_eq fun _ ↦ rfl
  have _ : IsScalarTower (center D) L D := .of_algebraMap_eq fun x ↦ by
```
The `letI` and `haveI` tactics play a different role in mathlib4, and here are not needed.

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,215 @@
+    rw [IsScalarTower.algebraMap_apply L (center D)]
+    congr
+    exact (equiv.apply_symm_apply x).symm
+  haveI : Algebra.IsAlgebraic (center D) D := .tower_top (K := L) _
```
```suggestion
  have _ : Algebra.IsAlgebraic (center D) D := .tower_top (K := L) _
```
See above

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,215 @@
+    ∃ x : D, x ∉ (⊥ : Subalgebra L D) ∧ IsSeparable L x := by
+  have ntrivial : center D ≠ ⊤ :=
+    congr(Subalgebra.toSubring $hcenter).trans_ne (Subalgebra.toSubring_injective.ne hneq)
+  set φ := Subalgebra.equivOfEq (⊥ : Subalgebra L D) (Subalgebra.center L D) hcenter.symm
```
```suggestion
  set φ := Subalgebra.equivOfEq (⊥ : Subalgebra L D) (.center L D) hcenter.symm
```

faenuccio reviewed (commented):
Thanks, this looks good to me! I am going to ask some maintainer to have a final look.

erdOne commented on Mathlib/Algebra/CharP/LinearMaps.lean:
```diff
@@ -0,0 +1,56 @@
+  element `x` such that `r • x = 0` implies `r = 0` (finding such element usually
+  depends on specific `•`), then the characteristic of `R` is equal to the
+  characteristic of the `R`-linear endomorphisms of `M`.-/
+theorem CharP_if {p : ℕ} [hchar : CharP R p]
```
I'm not sure if this is the right name. Maybe `Module.charP_end` or something similar.

Yu-Misaka replied:
Indeed that looks better, implemented, thanks!

erdOne commented on Mathlib/Algebra/CharP/Subring.lean:
```diff
@@ -33,4 +32,18 @@ instance subring (R : Type u) [Ring R] (p : ℕ) [CharP R p] (S : Subring R) : C
 
+/-- The characteristic of a division ring is equal to the characteristic
+  of its center-/
+theorem center_CharP_iff {R : Type u} [DivisionRing R] {p : ℕ} :
```
This should be called `charP_center_iff` and the direction of the iff should be flipped. Same for the one below.

mattrobball replied:
I think we can drop the double `charP` and also below

erdOne commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,215 @@
+
+open LinearMap
+
+local notation3 "k" => (Subring.center D)
```
I'm not entirely sure about this. I would think the mathliby approach is to have `[Algebra k D]` and a new typeclass called `IsCentral k D`. 

But if you have managed to restate all the results under that generality then it's fine.

Yu-Misaka replied:
Good point, I thought you might mean `exists_separable_mem_of_not_central'`, but `Algebra.IsCentral` seems more general than the `hcenter` in it... I'll try to generalize that a bit. Thanks!

Yu-Misaka replied:
Thanks! I've switched `exists_separable_mem_of_not_central'` to `IsCentral`.

erdOne commented:
Regarding the `large-import`: The file affected is a leaf file and the new import seems reasonable given the filename.

faenuccio commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,215 @@
+open LinearMap
+
+local notation3 "k" => (Subring.center D)
+private def δ : D → D →ₗ[k] D := mulLeft k - mulRight k
```
I take my share of responsibility here, I suggested that it should be private because it seems to me to be somewhat specific to the context. I agree that it could/shuold live with `mulLeft` and `mulRight`, provided we find a right name (`\delta` is probably not ok, I would go for `bra`: any idea, @mattrobball ?)

mattrobball replied:
It looks like the adjoint action of gl(V) on itself.

faenuccio replied:
Yes, this is what motivated me to call it `bra` for bracket. But perhaps @Yu-Misaka can try to use [LieDerivation.ad](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Lie/Derivation/AdjointAction.html#LieDerivation.ad) instead?

mattrobball replied:
Do we use `𝖆𝖉` currently?

mattrobball replied:
Oh! If we have something already, we should try to plug into that as much as possible.

Yu-Misaka replied:
Thanks! @faenuccio @mattrobball I've replaced all `δ` with `LieAlgebra.ad`. However I didn't manage to find existing lemma to replace `ad_iterate_succ` (originally private lemma `δ_iterate_succ`), so I placed it down the main theorem as a `have`.

faenuccio replied:
Great, I believe the situation of this PR is very good.

Yu-Misaka replied:
Sorry, I just noticed that `ad_iterate_succ` can be deduced directly from `Function.iterate_succ_apply'`...

Yu-Misaka commented on Mathlib/Algebra/CharP/LinearMaps.lean:
```diff
@@ -0,0 +1,56 @@
+  element `x` such that `r • x = 0` implies `r = 0` (finding such element usually
+  depends on specific `•`), then the characteristic of `R` is equal to the
+  characteristic of the `R`-linear endomorphisms of `M`.-/
+theorem CharP_if {p : ℕ} [hchar : CharP R p]
```
True, I'm thinking of namespace "Module" as @erdOne suggested changing it to `Module.charP_end`

mattrobball replied:
Did we settle a new name for this?

Yu-Misaka replied:
Yes, I've already placed it under namespace `Module` and renamed to `charP_end`.

Yu-Misaka commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,215 @@
+  | succ _ h => rw [Function.iterate_succ', Function.comp_apply, h,
+    pow_succ', mul_assoc, mulLeft_apply]
+
+private lemma r_pow (a : D) (n : ℕ) : ∀ x : D, ((mulRight k a) ^ n) x = x * (a ^ n) := by
```
Indeed, and both `r_pow` and `l_pow` are used only once so I've removed them.

Yu-Misaka commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,215 @@
+
+private lemma δ_def (a x : D) : δ a x = mulLeft k a x - mulRight k a x := rfl
+
+private lemma l_pow (a : D) (n : ℕ) : ∀ x : D, ((mulLeft k a) ^ n) x = (a ^ n) * x := by
```
They are indeed unnecessary. Thanks!

mattrobball commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,198 @@
+
+open Subring Algebra in
+/-- Jacobson-Noether theorem: For a non-commutative division algebra `D`
+  that is algebraic over a field `L`, if the center of
```
Is this really saying that `L` is the center? It seems more general a priori to me.

faenuccio replied:
I think this is precisely [Algebra.IsCentral](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Central/Defs.html#Algebra.IsCentral): it says that the center of `D` coincides with the image of `L` in `D`.

mattrobball replied:
But now I am confused by the naming here. I would think `exists_separable_of_not_top` or `exists_separable_of_not_comm` We really need to train a model to generate declaration names.

faenuccio replied:
I see what you mean, but mathematically we would simply identify `L` with its image, so saying that an element is not in `top` really means that it is not central and I feel that `not_central` (or `not_comm`, see below) conveys more mathematical meening. Concerning `not_comm`, I have the slight feeling that this would refer to a couple of elements commuting with each other rather than an element commuting with everything, but I agree this is not very sound.

mattrobball replied:
It seems strange to me that we have `IsCentral` as a parameter and we use `not_central` in the name. Also, what is `mem` doing? I couldn't quite parse it. It would nice to have some AI trained to generate these names from the declaration itself.

mattrobball replied:
Either way 

bors d=faenuccio

faenuccio replied:
@Yu-Misaka OK, we're at the end. Following Matt's advice, I would like that your rename the two main theorems by
1. Removing "mem" (it applies to conclusion of the form "something belongs to...", which is not the case)
2. Changing `central` to `isCentral`.

Then I will merge it.

faenuccio replied:
Thank you so much for your work!

Yu-Misaka replied:
Thank you for your help! It's been a pleasure working with you!

I've implemented the change;)

mattrobball replied:
Oh! I realized I was confused by the `of` since the **conclusion** is that you have a separable non central element.

mattrobball replied:
Right?

faenuccio replied:
Yep!

Yu-Misaka replied:
Sure, no problem;-)

mattrobball commented on Mathlib/FieldTheory/JacobsonNoether.lean:
```diff
@@ -0,0 +1,198 @@
+/-- Jacobson-Noether theorem: For a non-commutative division algebra
+  `D` that is algebraic over its center `k`, there exists an element
+  `x` of `D \ k` that is separable over `k`. -/
+theorem exists_separable_mem_of_not_central (H : k ≠ (⊤ : Subring D)) :
```
I would think `of_not_comm` is more fitting. Similarly below, the assumptions say something more than a prime.

Yu-Misaka commented on Mathlib/Algebra/GroupWithZero/Conj.lean:
```diff
@@ -20,3 +20,7 @@ variable {α : Type*} [GroupWithZero α] {a b : α}
   congr! 2 with c
   exact and_congr_right (mul_inv_eq_iff_eq_mul₀ · |>.symm)
+
+lemma conj_pow₀ {s : ℕ} {a d : α} (ha : a ≠ 0) : (a⁻¹ * d * a) ^ s = a⁻¹ * d ^ s * a :=
```
Indeed, I've placed them under namespace `GroupWithZero` for now (I'm not sure if this is the proper name though)

faenuccio commented:
Thanks!

faenuccio commented:
bors r+
